[BEAM-7021]: Added ToString transform on Python SDK

mszb opened the pull request:
Added ToString class in the combiners transformation on the Python SDK. Which includes three functions.
1. ToString.Kvs() : Converts KV Collection into string. Takes `delimiter` param as an optional field. (default delimiter is comma ",")
2. ToString.Elements() : Converts Element into string. Takes `delimiter` param as an optional field. (default delimiter is comma ",")
3. ToString.Iterables(): Converts Iterables objects into string. Takes `delimiter` param as an optional field. (default delimiter is comma ",")


Test cases also added for these methods.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

Commit:
[BEAM-7131] cache executable stage output to prevent re-computation

Commit:
Spark portable runner: cache all re-used RDDs by default

robertwb committed:
[BEAM-7364] Avoid possible signed integer overflow in hash.

tweise committed:
[BEAM-7348] Support environment expiration

Commit:
Make Jet Runner usable and documented

Commit:
[BEAM-7135] Spark: close job bundle factory

yifanzou committed:
[BEAM-7371] fix dependency check

mxm committed:
[BEAM-7366] Include the project's SourceSets in Spotless target

Spotless currently traverses only the project files to discover the source
files. This can causes problems when the project directory uses a source set
which spans outside the project folder. This is the case for the Flink Runner
projects which pull in their main source from `runners/flink` with the version
specific code being in `runners/flink/1.x`.

mxm committed:
Apply Spotless after fixing source directory detection

markflyhigh committed:
[BEAM-6908] Enable WordCountIT benchmark on Py36+37

tvalentyn committed:
Add a passing test to postcommit suites.

Commit:
[BEAM-7021]: added ToString transform on Python SDK

Commit:
[BEAM-7021]: added ToString transform on Python SDK

[BEAM-7021]: added ToString transform test on Python SDK

[BEAM-7021]: added comments on the python ToString transformation classes

Commit:
[BEAM-7021]: added ToString transform test on Python SDK

Commit:
Update stability critical jobs status

mxm committed:
Move StateContextsTest to let it execute as part of the tests

mxm committed:
[BEAM-7378] Fix Python PVR tests by adding a Kafka client dependency

The Expansion Server is started with the Flink Job Server and needs to have a
Kafka client to correctly expand transforms. The Kafka module was recently
changed to not bundle a Kafka client.

mxm committed:
Change scope of runtime dependency to runtime

mxm committed:
Merge pull request #8633: [BEAM-7378] Fix Python PVR tests by adding a Kafka client dependency

mxm committed:
[BEAM-7378] Add trigger to Python Flink PVR tests for Kafka module

This runs the cross-language Kafka tests in case of changes to the Kafka
module.

aromanenko-dev committed:
Merge pull request #8634: [BEAM-7378] Add trigger to Python Flink PVR tests for Kafka module

Commit:
[BEAM-7021]: added comments on the python ToString transformation classes

mszb commented:
R: @jibs 
R: @charithe 
R: @shehzaadn-vd

robertwb committed:
Merge pull request #8605 [BEAM-7135] Spark: close job bundle factory

iemejia committed:
[BEAM-7230] Make PoolableDataSourceProvider a static singleton
And simplifies all the inner method construction that was error-prone.

jbonofre committed:
Merge pull request #8635 from iemejia/BEAM-7230-jdbc-fix-pool-instantiation

[BEAM-7230] Make PoolableDataSourceProvider a static singleton

pabloem committed:
Merge pull request #8630 from Ardagan/MonthDash

Update stability critical jobs status

pabloem committed:
Merge pull request #8627 from tvalentyn/patch-51

[BEAM-7181] Unskip a passing test on Python 3.

TheNeuralBit committed:
[BEAM-6872] Add hook for user-defined JVM initialization in workers (#8104)

* Create helper method for ordered service loading

* [BEAM-6872] Add hook for user-defined JVM initialization in workers

Add BeamWorkerInitializer interface and logic to run it from the
Dataflow and portability worker harnesses.

* Add tests for worker initialization to Java SDK Harness

* !fixup fix spotless and other tests affected by the addition of an initializer

* !fixup add NonNull annotation to package-info

* !fixup add comment so people dont feel tempted to move the runOnStartup call

* !fixup make BeamWorkerInitializer an interface with default implementations

* !fixup rename to sdk.harness.JvmInitializer

* !fixup apply suggested comment changes

Co-Authored-By: Lukasz Cwik <lcwik@google.com>

* !fixup spotlessApply

markflyhigh committed:
Merge pull request #8628 from markflyhigh/py36-ws-benchmark

[BEAM-6908] Enable WordCountIT benchmark on Py36 and Py37

iemejia committed:
Make sfl4j runtime only for Samza Job Server too

yifanzou committed:
Merge pull request #8615 [BEAM-7371] fix dependency check

Commit:
Update .test-infra/jenkins/README

markflyhigh committed:
Merge pull request #8643 from markflyhigh/update-readme

Update .test-infra/jenkins/README

lukecwik committed:
Attempt to fix lots of URNs.

youngoli committed:
[BEAM-7383] Adding strict flag to runners to validate with vet runner

Creates a pipeline option called beam_strict that's supported by the
direct and universal runner. Uses the vet runner to perform the
verification for strict mode.

ttanay committed:
[BEAM-6695] Latest PTransform for Python SDK (#8206)

* [BEAM-6695] Latest PTransform for Python SDK

Added Latest PTransform and Combine Fns for the Python SDK.
Latest PTransform is used to compute the element(s) with the
latest timestamp from a PCollection.

udim committed:
Merge pull request #8593: [BEAM-7364] Avoid possible signed integer overflow in hash.

youngoli committed:
[BEAM-7354] Starcgen fix when no identifiers specified. (#8611)

If no identifiers are specified, the string.Split call as it is now
generates a slice of length 1 with one empty string, which isn't a valid
identifier and breaks the tool. This PR fixes it to the correct
behavior, to use an empty slice when there are no identifiers.

tvalentyn committed:
[BEAM-7365] Reduce the volume of test data in fastavro branch of test_dynamic_work_rebalancing_exhaustive to match the volume of avro branch.

iemejia committed:
[BEAM-6888] Enable SpotBugs in JdbcIO

jbonofre committed:
Merge pull request #8647 from iemejia/BEAM-6888-jdbc-enable-spotbugs

[BEAM-6888] Enable SpotBugs in JdbcIO

iemejia committed:
[BEAM-3934] Use try with resources to close the reader on JavaReadViaImpulse

mxm committed:
Merge pull request #8624: [BEAM-7366] Include the project's SourceSets in Spotless target

vectorijk committed:
[BEAM-5822] Vendor bytebuddy (#8357)

tvalentyn committed:
Support sync_interval in TestAvro._write_data() to simplify test scenarios.

chamikaramj committed:
Merge pull request #8646: [BEAM-7365] Reduces the volume of test data in fastavro branch to match the volume of avro branch

lukecwik committed:
Fix spotless errors

Commit:
refactor to use pushDataset everywhere

udim committed:
[BEAM-6429] side-input type coercion for multimaps

This might also resolve BEAM-4782.

Removes unused _input_element_coder, whose usage was removed in
PR #5302.

Includes squashed commits by @robertwb.

youngoli committed:
[BEAM-7393] Misc. improvements to Go error formatting.

Mainly two changes:
1. Remove default top-level error messages. If a top-level msg wasn't
   set then just don't display any.
2. Align the "Caused by:" msg on wrapped errors with the context
   messages, to improve readability.

udim committed:
[BEAM-4046] Allow running gradlew from repo subdirs

Example:
```sh
$ cd sdks/python
$ ... # use python tools
$ ../../gradlew lint
```

Also reduces verbosity of wrapper script.

udim committed:
Reduce verbosity of setup.py sdist and futurize

My poor eyes.

youngoli committed:
[BEAM-6880] Removing python integration test for Java ULR.

I missed this when removing ULR code, but nothing runs this test so
it wasn't causing problems.

udim committed:
[BEAM-6380] Allow rewinding PipeStream

Can only rewind to the beginning of the last block read.

Also fixes error detection in test thread.

robertwb committed:
Merge pull request #8658 [BEAM-6880] Removing python integration test for Java ULR.

mxm committed:
Merge pull request #8637: [BEAM-7305] Make Jet Runner usable and documented

iemejia committed:
Merge pull request #8642: Make sfl4j runtime only for Samza Job Server

ttanay committed:
[BEAM-7023] WithKeys PTransform for Python SDK

iemejia committed:
Refine runners/reference/java dependencies and fix minor static analysis issues

reuvenlax committed:
Merge pull request #8613: [BEAM-7356] Support generic field selections in Filter.

aaltay committed:
Merge pull request #8655 from youngoli/go-errors

[BEAM-7393] Misc. improvements to Go error formatting.

tweise committed:
Merge pull request #8599: [BEAM-7348] Support environment expiration

pabloem committed:
Merge pull request #8663 from ttanay/beam-7023

[BEAM-7023] WithKeys PTransform for Python SDK

tvalentyn committed:
[BEAM-7203] Signal Dataflow workers to use fastavro library on Python 3. (#8652)

lukecwik committed:
[BEAM-7409] Fix minor issues with \ and = in release guide.

davidyan74 committed:
[BEAM-7408] Make code example consistent with diagram

boyuanzz committed:
Merge pull request #8672 from lukecwik/release_guide

[BEAM-7409] Fix minor issues with \ and = in release guide.

youngoli committed:
[BEAM-7086] Refactoring repetitive error code.

Straightforward change, make some function literals for the really
repetitive errors.

fabito committed:
[BEAM-6027] Fix slow downloads when reading from GCS (#8553)

udim committed:
Merge pull request #8656: [BEAM-4046] Allow running gradlew from repo subdirs

chamikaramj committed:
Merge pull request #8661: [BEAM-6380] Allow rewinding PipeStream

udim committed:
Merge pull request #8657: Reduce verbosity of setup.py sdist and futurize

aaltay committed:
Merge pull request #8674 from youngoli/beam7086

[BEAM-7086] Refactoring repetitive error code.

Ardagan committed:
[BEAM-6284] Improve error message on waitUntilFinish. (#8629)

* Improve error message on waitUntilFinish.

Allow for infinite wait.

robertwb committed:
Merge pull request #8558 [BEAM-7131] Spark: cache output to prevent re-computation

shehzaadn-vd commented:
Looks good to me to create a public repo PR. @charithe @jibs your thoughts?

shehzaadn-vd reviewed (approved):
Looks good to me to create a public repo PR. Thanks @mszb

adude3141 committed:
[BEAM-4046] Allow running gradlew from repo subdirs (windows)

kaplanbora committed:
Fix command on flink runner docs

Flink application needs `run` command so it can start running.

iemejia committed:
Merge pull request #8665: Refine runners/reference/java dependencies and fix minor static analysis issues

melap committed:
Merge pull request #8673: [BEAM-7408] Make code example consistent with diagram

davidyan74 committed:
Merge pull request #8670: [BEAM-7408] Make custom options example consistent across languages

mxm committed:
Merge pull request #8678: [BEAM-7417] Fix incorrect command on flink runner docs

coheigea committed:
Replace assertEquals(boolean, xyz) with assertBoolean(xyz) (#8037)

adude3141 committed:
Merge pull request #8676: [BEAM-4046] Allow running gradlew from repo subdirs (windows)

angoenka committed:
[BEAM-7422] Create new java worker in flink python test

ananvay committed:
splitting support over FnAPI for Go SDK

Hannah-Jiang committed:
[BEAM-6693] ApproximateUnique transforms for Python SDK (#8535)

* [BEAM-6693]ApproximateUnique transforms for Python SDK

kennknowles committed:
[BEAM-7406] Include slf4j-jdk14 in Dataflow worker uber jar

lukecwik committed:
Update Python Dataflow precommit to use worker jar built locally

angoenka committed:
Merge pull request #8682 from angoenka/fix_expansion_task

[BEAM-7422] Create new java worker in flink python test

ananvay committed:
splitting support over FnAPI for Go SDK. updated

udim committed:
Merge pull request #8654: [BEAM-6429] side-input type coercion for multimaps

Commit:
Update datasource.go

lukecwik committed:
[BEAM-7387] Attempt to fix lots of URNs to follow common pattern.

udim committed:
Merge pull request #8683: [BEAM-5791] splitting support over FnAPI for Go SDK

lostluck committed:
Fix small typo in vet.go

iemejia committed:
Merge pull request #8648: [BEAM-3934] Use try with resources to close the reader on JavaReadViaImpulse

alexvanboxel committed:
[BEAM-7426] Add underscore support on Field lexer

Underscore is a common used word delimiter in field names, the current
FieldSpecifierNotationLexer only support alpha-numeric values for field
name character.

The upcoming Protobuf schema support will emit underscores in the field
names, so field names should support underscore.

aaltay committed:
Merge pull request #8685 from lostluck/patch-4

Fix small typo in vet.go

aaltay committed:
Merge pull request #8644 from youngoli/beam7383

[BEAM-7383] Adding strict flag to runners to validate with vet runner

iemejia committed:
Refactor default bundle size into a constant in TextIO

Make more readable default bundle size in JavaReadViaImpulse

mxm committed:
[BEAM-7421] Add missing transform payload translator for Reshuffle

Reshuffle is overwritten in various Runners (e.g. Flink, Spark) but the
translation code does not kick-in due to a missing TransformPayloadTranslator
alongside with a TransformPayloadTranslatorRegistrar.

iemejia committed:
Merge pull request #8692: Refactor default bundle size into a constant in TextIO

mxm committed:
[cleanup] Remove dead code from Flink Runner

This removes left-over code which does not serve a purpose anymore.

mxm committed:
Merge pull request #8694: [cleanup] Remove dead code from Flink Runner

mxm committed:
Merge pull request #8693: [BEAM-7421] Add missing transform payload translator for Reshuffle

The composite transform (Reshuffle) is not changed, only its resulting Flink/Spark translation is now as specified in the Runners' translator classes. For non-portable pipelines, the pure existence of the translator will cause Reshuffle to be recognized as a translatable transform. For portable pipelines, it will set the FunctionSpec field of the transform, while preserving the subtransforms.

Note that this only affects the Java SDK.

reuvenlax committed:
Merge pull request #8689: [BEAM-7426] Add underscore support on Field lexer

rakeshcusat committed:
[BEAM-7141] Add window & timestamp in timer callback method argument (#8408)

Commit:
Merge remote-tracking branch 'origin/BEAM-7021' into BEAM-7021

Commit:
[BEAM-7021] fix a typo and resolve linting issues.
